include/zephyr/sys/util.h

```diff
@@ -26,6 +26,90 @@
 #include <zephyr/types.h>
 #include <stddef.h>
 
+/** @brief Number of bits that make up a type */
+#define NUM_BITS(t) (sizeof(t) * 8)
+
+#ifdef __cplusplus
+template <typename T> static inline constexpr int __z_log2_impl(T x)
+{
+	if (sizeof(x) < sizeof(unsigned int)) {
+		return ((x) < 1) * (-1) +
+		       ((x) >= 1) * (NUM_BITS(unsigned int) - __builtin_clz(x) - 1);
+	} else if (sizeof(x) <= sizeof(unsigned long long)) {
+		return ((x) < 1) * (-1) +
+		       ((x) >= 1) * (NUM_BITS(unsigned long long) - __builtin_clzll(x) - 1);
+	}
+
+	/* No declaration of __ASSERT is available here */
+	static_assert(sizeof(x) <= sizeof(unsigned long long), "unsupported type for LOG2()");
+
+	return -1;
+}
+
+template <typename T> static inline constexpr int __z_log2ceil_impl(T x)
+{
+	if (sizeof(x) < sizeof(unsigned int)) {
+		return (x > 1) * (NUM_BITS(unsigned int) - __builtin_clz(x - 1));
+	} else if (sizeof(x) <= sizeof(unsigned long long)) {
+		return (x > 1) * (NUM_BITS(unsigned long long) - __builtin_clzll(x - 1));
+	}
+
+	/* No declaration of __ASSERT is available here */
+	static_assert(sizeof(x) <= sizeof(unsigned long long), "unsupported type for LOG2CEIL()");
+
+	return -1;
+}
+
+template <typename T> static inline constexpr uint64_t __z_nhpot_impl(T x)
+{
+	int l2c = __z_log2ceil_impl(x);
+
+	return (l2c != NUM_BITS(unsigned long long)) * (1ULL << l2c);
+}
+#else
+#define __z_log2_impl(x)                                                                           \
+	(((x) < 1) * (-1) +                                                                        \
+	 ((x) >= 1) * _Generic((x), char                                                           \
+			       : (NUM_BITS(unsigned int) - __builtin_clz(x) - 1), unsigned char    \
```

Comment:
Is omitting `signed char` intentional?

Reply:
@jwakely - no... because .. it wasn't omitted?

Reply:
Thanks for clarifying - patches welcome!